app/javascript/src/Users/components/PermissionsForm.jsx

```diff
@@ -0,0 +1,109 @@
+// @flow
+
+import 'raf/polyfill'
+import 'core-js/es6/map'
+import 'core-js/es6/set'
+import 'core-js/es6/array'
+
+import React from 'react'
+
+import { Inputs } from 'Users/components/Inputs'
+import { RoleInput } from 'Users/components/RoleInput'
+import { UserRole } from 'Users/components/UserRole'
+import { FeatureAccessInput } from 'Users/components/FeatureAccessInput'
+import { FeatureAccess } from 'Users/components/FeatureAccess'
+import { ServiceFeatureAccess } from 'Users/components/ServiceFeatureAccess'
+import { ServiceAccessList } from 'Users/components/ServiceAccessList'
+import { ServiceAccess } from 'Users/components/ServiceAccess'
+
+import { isServicePermissionsGranted, getFeatureName, toggleElementInCollection } from 'Users/utils'
+import { createReactWrapper } from 'utilities/createReactWrapper'
+
+import type { Role, Feature } from 'Users/types'
+import type { Service } from 'Types'
+
+type Props = {
+  initialState: any,
+  features: Feature[],
+  services: Service[]
+}
+
+type State = {
+  role: Role,
+  checkedFeatures: Array<Feature | 'services'>,
```

Comment:
This looks weird, having an `Array` of `Feature` or `string services`?

Reply:
I agree it looks weird but this is how our API is defined. The array contains all *checked features* plus it may contain the term *services* that is not a feature. When it is _not_ present, it means the whole service set is enabled. When it _is_ present, it means services have been enabled manually and some may not been included. See following pictures:

![Screen Shot 2019-04-02 at 16 55 02](https://user-images.githubusercontent.com/11672286/55413330-6ec76880-5569-11e9-949c-4327b9dc7603.png)
>`admin_sections` contains only features (billing, settings, analytics, etc) and all services are enabled. `member_permission_service_ids` is `null`.

![Screen Shot 2019-04-02 at 16 55 36](https://user-images.githubusercontent.com/11672286/55413336-725aef80-5569-11e9-85b0-d45bf9ddcd64.png)
>`admin_sections` here contains the term `services` that indicates a set of services is specified in `member_permissions_service_ids`.

Reply:
What about defining a type `AdminSections` where it will be a list of `string` and then a `UserAdminSections a` where `a` would be a List `AdminSections`. Also, you could keep a type `Services` in your state among the role and serviceCheckedIds. Then `Features` defined somewhere (you don't care about them in the state of your app). Then you could decide if a Feature is checked based on the admiSections you've got from your initialState. That way would be type bulletproof